app/models/concerns/module_preferences_concern.rb

```diff
@@ -0,0 +1,53 @@
+module ModulePreferencesConcern
+  extend ActiveSupport::Concern
+
+  included do
+    has_many :module_preference_objects, as: :target, dependent: :destroy
+    accepts_nested_attributes_for :module_preference_objects
+
+    if self == Course
```

Comment:
Any better way to do this ? I've tried few different methods actually.

Basically, Course and instance need to include this concern. this function is different for them, other functions are same.